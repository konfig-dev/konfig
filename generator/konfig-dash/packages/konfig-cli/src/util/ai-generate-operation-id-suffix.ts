import { Operation, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOperationIdSuffix({
  prefix,
  path,
  method,
  operation,
  operations,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  operations: OperationObject[]
}): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })

  const existingIds = operations.map((o) => o.operation.operationId)
  const existingIdsStr = existingIds.join(', ').slice(0, -2)
  const systemPrompt = `You are a generator of operation IDs for an OpenAPI specification. You generate operation IDs in json format: "{ prefix: ... , suffix: ...}"`
  const prompt = generatePrompt({
    prefix,
    path,
    method,
    operation,
    existingIdsStr,
  })

  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.suffix === undefined || responseJson.suffix === null)
    throw Error('OpenAI returned an invalid json: it does not contain a suffix')

  // Post-process to ensure that rules are followed
  let suffix = responseJson.suffix.replace('_', '')
  suffix = suffix.charAt(0).toLowerCase() + suffix.slice(1)
  suffix = suffix.replace('retrieve', 'get')
  suffix = suffix.replace('fetch', 'get')
  console.log(`AI-generated operation id: ${prefix}${suffix}`)
  return suffix
}

function generatePrompt({
  prefix,
  path,
  method,
  operation,
  existingIdsStr,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  existingIdsStr: string
}): string {
  return `Generate an operation ID for this operation in my OpenAPI specification:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}
Description: ${operation.description}

The operation ID is in the format "prefix_suffix". The prefix is: ${prefix}. Generate the suffix and respond in json format.
The operation ID MUST follow ALL of these rules:

1. The suffix must be at least 3 characters long.
2. The suffix must be in camelCase and CANNOT include any underscores.
3. The suffix must not redundantly re-use words from the prefix. (For example: Store_get is preferred over Store_getStore.)
4. If the suffix contains both a verb and a noun, the noun should generally follow the verb. (For example, Store_placeOrder is better than Store_orderPlace).
5. Prefer simple verbs like "get", "create", "update", "delete", and "list" when possible.

Existing operation ids for this OpenAPI specification are as follows: ${existingIdsStr}.
The operation ID MUST be unique within this list of existing operation IDs. You can also use them as a reference to generate this operation id.`
}
