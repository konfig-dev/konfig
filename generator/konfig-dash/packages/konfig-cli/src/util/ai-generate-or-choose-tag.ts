import { Operation, TagObject, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOrChooseTag(
  operation: Operation,
  path: string,
  method: string,
  tags: TagObject[]
): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })
  const tagNames = tags.map((t) => t.name)

  // First, try and choose a tag programmatically without using OpenAI
  const tag = programaticallyChooseTag(operation, path, tagNames)
  if (tag !== undefined) return tag
  const systemPrompt = `You are a generator of tags for an OpenAPI specification. You generate tags in json format: "{ tag: ... }"`
  const prompt = generatePrompt(path, method, operation, tagNames)
  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.tag === undefined || responseJson.tag === null)
    throw Error('OpenAI returned an invalid json: it does not contain a tag')

  tags.push({ name: responseJson.tag })
  return responseJson.tag
}

function programaticallyChooseTag(
  operation: Operation,
  path: string,
  tagNames: string[]
): string | undefined {
  for (const tag of tagNames) {
    if (path.toLowerCase().includes(tag.toLowerCase())) return tag
    if (
      operation.summary &&
      operation.summary.toLowerCase().includes(tag.toLowerCase())
    )
      return tag
  }
  return undefined
}

function generatePrompt(
  path: string,
  method: string,
  operation: Operation,
  tagNames: string[]
): string {
  const tagNamesStr = tagNames.join(', ').slice(0, -2)
  return `The following operation from my OpenAPI specification is missing a tag:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}

Based on this information, choose the best tag from this list of existing tags on the OpenApi Specification: ${tagNamesStr}.
IF AND ONLY IF none of these existing tags are suitable, then generate a new tag. The new tag MUST be a single word, MUST be a noun, and MUST be singular. It MUST NOT be a combination of multiple words.

In general, a good tag name is related to the type of data being operated on (such as "Student"), or the type of operation being performed (such as "Authentication").

Respond with just the tag in json format: "{ tag: ... }" with no other information.`
}
