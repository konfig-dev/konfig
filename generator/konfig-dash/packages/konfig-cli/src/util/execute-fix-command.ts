import * as fs from 'fs-extra'
import {
  isJsonString,
  OBJECT_WITH_NO_PROPERTIES_RULE_NAME,
  parseSpec,
  filterSpecPaths,
  POTENTIAL_INCORRECT_DATA_TYPE_RULE_NAME,
} from 'konfig-lib'
import boxen from 'boxen'
import { Progress } from './fix-progress'
import { oasYamlDump } from './oas-yaml-dump'
import { fixOas } from './fix-oas'
import path from 'path'
import { oasJsonDump } from './oas-json-dump'
import { parseKonfigYaml } from './parse-konfig-yaml'

interface FixOptions {
  format?: boolean
  specInputPath?: string
  spec?: string
  auto?: boolean
  konfigDir?: string
  skipMissingResponseDescriptionFix?: boolean
  skipListUsageSecurity?: boolean
  alwaysYes?: boolean
  ci?: boolean
}

export async function executeFixCommand(options: FixOptions): Promise<void> {
  const flags = {
    ...options,
    auto: options.auto ?? true,
    alwaysYes: !!(options.alwaysYes == null || options.alwaysYes || options.ci), // ci mode should always confirm with yes
    ci: options.ci ?? false,
  }

  const { parsedKonfigYaml } = parseKonfigYaml({
    configDir: flags.konfigDir ?? process.cwd(),
  })

  // We only want to run fix in CI if the konfig.yaml has a specInputPath
  if (flags.ci && parsedKonfigYaml.specInputPath === undefined) {
    console.log(`No specInputPath found in konfig.yaml, skipping fix`)
    return
  }

  if (flags.spec === undefined) {
    flags.spec = parsedKonfigYaml.specPath
    flags.specInputPath = parsedKonfigYaml.specInputPath ?? flags.spec
  }

  if (flags.spec === undefined)
    throw Error(
      `Either specify path to OAS with -s flag or assign "specPath" field in "konfig.yaml"`
    )

  if (flags.specInputPath === undefined) throw Error(`This shouldn't happen`)

  const prependKonfigDir = ({ specPath }: { specPath: string }) => {
    return flags.konfigDir !== undefined
      ? path.join(flags.konfigDir, specPath)
      : specPath
  }

  const specInputPath = prependKonfigDir({ specPath: flags.specInputPath })
  const specOutputPath = prependKonfigDir({ specPath: flags.spec })
  const rawSpec = fs.readFileSync(specInputPath, 'utf-8')
  let spec = await parseSpec(rawSpec)

  // if konfig yaml filters any paths from the spec, remove them now
  if (parsedKonfigYaml.filterPaths) {
    filterSpecPaths({ spec: spec.spec, filter: parsedKonfigYaml.filterPaths })
  }

  if (flags.format) {
    fs.writeFileSync(flags.spec, oasYamlDump(spec))
    return
  }

  const progress: Progress = Progress.getSaved({
    konfigDir: flags.konfigDir ?? process.cwd(),
    progressYamlPathOverride: parsedKonfigYaml.progressYamlPath,
  })

  const {
    issuesFixed,
    infoDescriptionFixed,
    numberOfMissingResponseDescriptionsAdded,
    numberOfExamplesFixed,
    numberOfUpdatedOperationIds,
    numberOfMissingTags,
    numberOfTrailingSlashesFixed,
    numberOfMissing200ResponsesAdded,
    numberOfParametersConvertedToSecurityRequirements,
    numberOfEmptyRequestBodiesRemoved,
    numberOfListUsagesOfSecurityFixed,
    numberOfImproperlyNamedTags,
    numberOfRedundantSecurityAndParametersFixed,
    numberOfObjectsWithNoPropertiesFixed,
    numberOfSchemasDefined,
    numberOfParametersRemovedForNewSecurityScheme,
    numberOfUnstructuredRequestBodiesFixed,
    numberOfIgnorePotentialIncorrectTypeAdded,
    numberOfCustomModifications,
    numberOfIgnoreObjectsWithNoPropertiesAdded,
    numberOfEmptyResponseBodySchemasFixed,
    numberOfDuplicateTagNamesFixed,
    numberOfInvalidServerUrlsFixed,
    numberOfRedundantSecuritySchemesRemoved,
    numberOfExampleAndExamplesFixed,
    numberOfDisallowedHeaderNamesRemoved,
    numberOfNewTagNames,
    numberOfAnyOfTypeNullUsagesReverted,
    numberOfOas31UsagesFixed,
    numberOfExamplesUsageRemoved,
  } = await fixOas({
    spec,
    progress,
    konfigYaml: parsedKonfigYaml,
    alwaysYes: flags.alwaysYes,
    auto: flags.auto,
    ci: flags.ci,
    skipMissingResponseDescription: flags.skipMissingResponseDescriptionFix,
    skipFixListUsageSecurity: flags.skipListUsageSecurity,
  })

  fs.writeFileSync(
    specOutputPath,
    isJsonString(rawSpec) ? oasJsonDump(spec.spec) : oasYamlDump(spec.spec)
  )

  console.log(
    boxen(
      `Updated ${numberOfUpdatedOperationIds} operation IDs
Made ${numberOfCustomModifications} custom modifications
Renamed ${numberOfImproperlyNamedTags} improperly named tags
Removed ${numberOfDisallowedHeaderNamesRemoved} disallowed headers removed
Added ${numberOfIgnoreObjectsWithNoPropertiesAdded} ${OBJECT_WITH_NO_PROPERTIES_RULE_NAME} ignore rule
Added ${numberOfIgnorePotentialIncorrectTypeAdded} ${POTENTIAL_INCORRECT_DATA_TYPE_RULE_NAME} ignore rule
Added ${infoDescriptionFixed} missing info description
Added ${numberOfMissingResponseDescriptionsAdded} missing response descriptions
Added ${numberOfMissingTags} missing tags
Added ${numberOfEmptyResponseBodySchemasFixed} empty response body schemas
Added ${numberOfSchemasDefined} named schemas
Added ${numberOfMissing200ResponsesAdded} missing 2xx responses
Added ${numberOfNewTagNames} new tag names
Fixed ${numberOfDuplicateTagNamesFixed} duplicate tag names fixed
Fixed ${numberOfInvalidServerUrlsFixed} invalid server URLs fixed
Fixed ${numberOfUnstructuredRequestBodiesFixed} unstructured request bodies
Fixed ${numberOfObjectsWithNoPropertiesFixed} objects with no properties
Fixed ${numberOfExampleAndExamplesFixed} redundant "example" & "examples" fields
Fixed ${numberOfExamplesFixed} examples with invalid schemas
Fixed ${numberOfParametersConvertedToSecurityRequirements} parameters that should be security requirements
Fixed ${numberOfListUsagesOfSecurityFixed} list usages of security
Removed ${numberOfTrailingSlashesFixed} trailing slashes
Removed ${numberOfRedundantSecurityAndParametersFixed} redundant security requirement and parameters
Removed ${numberOfRedundantSecuritySchemesRemoved} redundant security schemes
Removed ${numberOfParametersRemovedForNewSecurityScheme} parameters replace with security requirement
Removed ${numberOfEmptyRequestBodiesRemoved} empty request bodies
Fixed ${numberOfAnyOfTypeNullUsagesReverted} anyOf type null usages
Fixed ${numberOfOas31UsagesFixed} OAS 3.1.x usages
Fixed ${numberOfExamplesUsageRemoved} "examples" usage`,
      {
        title: `Fixed ${issuesFixed} Issues`,
        borderColor: issuesFixed === 0 ? 'gray' : 'green',
        padding: 1,
      }
    )
  )
}
