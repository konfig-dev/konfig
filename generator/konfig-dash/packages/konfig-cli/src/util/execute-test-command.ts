import { CliUx } from '@oclif/core'
import execa, { ExecaSyncReturnValue, SyncOptions } from 'execa'
import {
  KonfigYamlGeneratorNames,
  KonfigYamlGeneratorConfig,
  RequiredEnvironmentVariablesConfig,
  TestConfig,
  KonfigYamlAdditionalGeneratorConfig,
} from 'konfig-lib'
import waiton from 'wait-on'
import kill from 'konfig-kill-port'
import path from 'path'
import { parseKonfigYaml } from './parse-konfig-yaml'
import { parseFilterFlag } from './parseFilterFlag'

const validateRequiredEnvironmentVariables = (
  config: RequiredEnvironmentVariablesConfig
) => {
  if (config.requiredEnvironmentVariables === undefined) return
  for (const environmentVariable of config.requiredEnvironmentVariables) {
    if (!(environmentVariable in process.env))
      CliUx.ux.error(
        `Missing required environment variable: ${environmentVariable}`
      )
  }
}

export async function executeTestCommand({
  filterInput,
  sequence,
}: {
  filterInput?: string
  sequence: boolean
}) {
  const filter = parseFilterFlag(filterInput)
  const configDir = process.cwd()
  const { generators, common, additionalGenerators } = parseKonfigYaml({
    configDir,
  })

  // kill any existing process on 4010
  const port = 4010
  CliUx.ux.log(`Killing any process using port ${port}`)
  try {
    await kill(port)
  } catch (e) {
    if (e instanceof Error) {
      if (e.message !== 'No process running on port') throw e
    }
  }

  // spawn process that run "konfig mock -d {specPath}" from konfig.yaml
  CliUx.ux.log('ðŸ’» Starting mock server')
  // TODO: ENG-1099 Use a function call here instead of CLI
  execa.command(`konfig mock -d ${common.specPath}`, {
    cwd: configDir,
    stdio: 'inherit',
    shell: true,
  })

  // if this process exits in any way, kill the mock server
  const handleTermination = async () => {
    CliUx.ux.log('ðŸ›‘ Killing mock server')
    await kill(port)
    process.exit()
  };

  process.on('exit', handleTermination)
  process.on('SIGINT', handleTermination)
  process.on('SIGTERM', handleTermination)
  process.on('SIGUSR1', handleTermination)
  process.on('SIGUSR2', handleTermination)

  await waiton({
    resources: [`http://127.0.0.1:${port}`],
    timeout: 60_000,
    validateStatus: (status) => {
      console.log(status)
      return status === 405 || status === 404 || status === 200
    },
  })
  CliUx.ux.log('âœ… Started mock server')

  validateRequiredEnvironmentVariables(common)
  const results: {
    generatorName: string
    result: ExecaSyncReturnValue<string>
  }[] = []
  const asyncResults: Promise<{
    generatorName: string
    result: ExecaSyncReturnValue<string>
  }>[] = []

  if (!sequence) CliUx.ux.log('Running tests in parallel')
  else CliUx.ux.log('Running tests sequentially')

  function runTest({
    test,
    outputDirectory,
    disabled,
    generatorName,
    generatorConfig,
  }: {
    test?: TestConfig['test']
    outputDirectory?: string
    disabled?: boolean
    generatorName: string
    generatorConfig: KonfigYamlGeneratorConfig
  }) {
    if (filter !== null && !filter.includes(generatorName)) return
    if (disabled) return
    const testScript =
      test?.script === undefined
        ? defaultTestScript({
            generatorName: generatorName as KonfigYamlGeneratorNames,
            generatorConfig,
          })
        : test.script
    if (outputDirectory === undefined)
      throw Error(
        `Generator "${generatorName}" must provide "outputDirectory" for "testScript" to work`
      )
    if (sequence)
      CliUx.ux.action.start(`Running tests for ${generatorName} SDK`)

    // verify environment variables are set if specified
    if (test !== undefined) validateRequiredEnvironmentVariables(test)

    const commandOptions: SyncOptions = {
      cwd: path.join(configDir, outputDirectory),
      stdio: 'inherit',
      shell: true,
    }

    if (sequence) {
      // Sequential
      for (const idx in testScript) {
        const result = execa.commandSync(testScript[idx], commandOptions)
        if (parseInt(idx) == testScript.length - 1) {
          results.push({
            generatorName,
            result,
          })
        }
      }
    } else {
      // Parallel
      asyncResults.push(
        new Promise(async (resolve) => {
          for (const idx in testScript) {
            const result = await execa.command(testScript[idx], commandOptions)
            if (parseInt(idx) === testScript.length - 1) {
              resolve({ generatorName, result })
            }
          }
        })
      )
    }
    if (sequence) CliUx.ux.action.stop()
  }

  for (const [generatorName, generatorConfig] of [
    ...Object.entries(generators),
    ...(additionalGenerators ? Object.entries(additionalGenerators) : []),
  ]) {
    runTest({ ...generatorConfig, generatorName, generatorConfig })
  }

  results.push(...(await Promise.all(asyncResults)))

  if (results.length === 0) {
    CliUx.ux.error('No tests were run')
  }

  let anyFailed: boolean = false
  for (const { result, generatorName } of results) {
    if (!result.failed) continue
    CliUx.ux.warn(`tests for ${generatorName} failed`)
  }
  if (anyFailed) CliUx.ux.error('Test failure occurred')

  // If we made it here then we successfully ran all tests
  CliUx.ux.info('Successfully ran all tests!')

  process.exit()
}

const defaultTestScripts: Record<
  KonfigYamlGeneratorNames,
  ((config: KonfigYamlGeneratorConfig) => string[]) | undefined
> = {
  typescript: () => ['yarn', 'yarn test', 'yarn build'],
  csharp: () => ['dotnet test'],
  java: () => ['mvn test'],
  ruby: () => [
    // If you don't remove .gem files you get:
    // You have one or more invalid gemspecs that need to be fixed.
    // The gemspec at snaptrade-sdks/sdks/ruby/snaptrade.gemspec is not valid. Please fix this gemspec.
    // The validation error was 'snaptrade-1.0.0 contains itself (snaptrade-1.0.0.gem), check your files list'
    `rm *.gem || true`, // "|| true" is used to ensure command exits w/o code of 1 (https://superuser.com/a/887349),
    'bundle install',
    `bundle exec rspec`,
  ],
  python: (config) => {
    if (!('packageName' in config))
      throw Error('Got unexpected config for Python SDK')
    return [
      'poetry install',
      `poetry run pytest --cov=${config.packageName} -o cache_dir=${path.join(
        config.outputDirectory,
        '.pytest_cache'
      )}`,
    ]
  },
  php: () => ['composer install', path.join('.', 'vendor', 'bin', 'phpunit')],
  go: () => ['go clean -testcache', 'go test ./... -v'],
  swift: () => ['swift test'],
  android: () => {
    throw Error('Android generator not supported, use Java instead')
  },
  kotlin: () => {
    throw Error('Kotlin generator not supported, use Java instead')
  },
  objc: () => {
    throw Error('Objective-C generator not supported, use Swift instead')
  },
}
function defaultTestScript({
  generatorName,
  generatorConfig,
}: {
  generatorName: KonfigYamlGeneratorNames
  generatorConfig:
    | KonfigYamlGeneratorConfig
    | KonfigYamlAdditionalGeneratorConfig
}): string[] {
  const name =
    'generator' in generatorConfig ? generatorConfig.generator : generatorName
  const testScript = defaultTestScripts[name]
  if (testScript === undefined)
    throw Error(
      `Could not find default test script for generator with name: "${generatorName}"`
    )
  return testScript(generatorConfig)
}
