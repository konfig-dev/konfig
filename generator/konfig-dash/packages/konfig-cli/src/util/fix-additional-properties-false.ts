import { Spec, recurseObject } from 'konfig-lib'

/**
 * "additionalProperties: false" causes issue when mocking an API.  The
 * following type of errors will appear for no reason: "âœ–  error     Request
 * body must NOT have additional properties; found 'userId'" but userId is
 * actually a property in the schema so the error is just wrong.  To get around
 * this I added this rule to remove "additionalProperties: false" from the
 * schema.
 */
export function fixAdditionalPropertiesFalse({ spec }: { spec: Spec }): number {
  let numberOfAdditionalPropertiesFixed = 0
  recurseObject(spec.spec, ({ value: schema }) => {
    if (typeof schema !== 'object') return
    if (schema === null) return
    if (schema.type !== 'object') return
    if (schema.additionalProperties === undefined) return
    if (schema.additionalProperties === true) return
    delete schema.additionalProperties
    numberOfAdditionalPropertiesFixed++
  })
  return numberOfAdditionalPropertiesFixed
}
