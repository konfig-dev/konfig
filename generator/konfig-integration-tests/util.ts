import { execa } from "execa";
import * as path from "path";
import yaml from "js-yaml";
import { expect } from "vitest";
import * as fs from "fs";
import {
  KONFIG_YAML_NAME,
  KonfigYaml,
  KonfigYamlAdditionalGeneratorConfig,
  KonfigYamlGeneratorConfig,
  KonfigYamlType,
} from "konfig-lib";
import waitOn from "wait-on";

const USE_RUN =
  process.env.GITHUB_ACTIONS === "true" ||
  (process.env.RUN && process.env.RUN.trim() !== "");

console.log("USE_RUN", USE_RUN);

// relative path from this file to: ""../konfig-dash/packages/konfig-cli/bin/dev""
// use nodejs __dirname to get the current directory of this file
const KONFIG_CLI_PATH = path.join(
  __dirname,
  `../konfig-dash/packages/konfig-cli/bin/${USE_RUN ? "run" : "dev"}`
);

/**
 * parse 'python-pydantic-empty-response' from 'tests/python-pydantic-empty-response.test.ts > python-pydantic-empty-response'
 */
function getCurrentTestName(): string {
  const currentTestName = expect.getState().currentTestName;
  const parsed = currentTestName?.split(" > ")[1];
  if (!parsed) {
    throw new Error(`Unable to parse current test name: ${currentTestName}`);
  }
  return parsed;
}

export async function e2e(
  mockServerPort: number,
  customAssertions?: () => void
) {
  if (!getCurrentTestName()) {
    throw new Error("Unable to get current test name");
  }
  const sdkDir = path.join(__dirname, "sdks", getCurrentTestName());

  const apiUrl = process.env.KONFIG_API_URL ?? "http://127.0.0.1:8911";

  // wait until KONFIG_API_URL/healthz is available
  await waitOn({
    resources: [`${apiUrl}/healthz`],
    timeout: 100000,
  });

  const env = {
    // This ensure oclif doesn't bug out
    // See: https://github.com/oclif/oclif/issues/1161#issuecomment-1661372245
    NODE_ENV: "",
  };

  // run "konfig generate -d -I" inside the path
  await execa(KONFIG_CLI_PATH, ["generate", "-d", "-I"], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // run "konfig test" inside the path
  await execa(KONFIG_CLI_PATH, ["test", "-p", mockServerPort.toString()], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // validate top-level README.md
  const konfigYamlPath = path.join(sdkDir, KONFIG_YAML_NAME);
  if (!fs.existsSync(konfigYamlPath))
    throw new Error(
      `Make sure you are in same directory as '${KONFIG_YAML_NAME}'`
    );
  const konfigYamlRaw = fs.readFileSync(konfigYamlPath, "utf-8");
  const loadedKonfigYaml: KonfigYamlType = yaml.load(
    konfigYamlRaw
  ) as KonfigYamlType;
  const parsedKonfigYaml = KonfigYaml.parse(loadedKonfigYaml);

  const generators: (
    | KonfigYamlGeneratorConfig
    | KonfigYamlAdditionalGeneratorConfig
  )[] = [
    ...Object.values(parsedKonfigYaml.generators),
    ...(parsedKonfigYaml.additionalGenerators !== undefined
      ? Object.values(parsedKonfigYaml.additionalGenerators)
      : []),
  ];
  for (const generator of generators) {
    expect(
      normalizeDocumentation(
        fs.readFileSync(
          path.join(sdkDir, generator.outputDirectory, "README.md"),
          "utf-8"
        )
      )
    ).toMatchSnapshot();
  }

  if (customAssertions) {
    customAssertions();
  }
}

// Removes the [GitHub last commit] line from the README
function normalizeDocumentation(readme: string) {
  // matches [GitHub last commit] to the end of the line
  const pattern = /\[!\[GitHub last commit\].*$\n?/gm;
  return readme.replace(pattern, "");
}
