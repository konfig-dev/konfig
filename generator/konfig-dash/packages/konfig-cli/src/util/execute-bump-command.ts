import { CliUx } from '@oclif/core'
import { getPackageVersion, setPackageVersion, KonfigYamlType } from 'konfig-lib'
import { generateKonfigYamlString } from '../util/generate-konfig-yaml-string'
import { incrementVersionWithPreservedTag } from '../util/increment-version-with-preserved-tag'
import semver from 'semver'
import * as fs from 'fs-extra'

export async function executeBumpCommand({
  filter,
  releaseType,
  loadedKonfigYaml,
  konfigYamlPath,
}: {
  filter: string[] | null
  releaseType: semver.ReleaseType
  loadedKonfigYaml: KonfigYamlType
  konfigYamlPath: string
}): Promise<void> {

  if (filter !== null)
    CliUx.ux.info(`Performing ${releaseType} bump of ${filter.join(', ')}`)
  else CliUx.ux.info(`Performing "${releaseType}" bump`)

  for (const [generatorName, generatorConfig] of [
    ...Object.entries(loadedKonfigYaml.generators),
    ...Object.entries(loadedKonfigYaml.additionalGenerators ?? []),
  ]) {
    if (filter !== null && !filter.includes(generatorName)) continue
    if ('disabled' in generatorConfig && generatorConfig.disabled) continue
    const currentVersion = getPackageVersion({ generatorConfig })
    const newVersion = incrementVersionWithPreservedTag(currentVersion, releaseType)
    CliUx.ux.info(`Bumping ${generatorName} SDK from ${currentVersion} --> ${newVersion}`)
    if (newVersion === null)
      CliUx.ux.error(`Could not compute new version for ${generatorName} SDK currently at ${currentVersion}`)
    setPackageVersion({ generatorConfig, version: newVersion })
  }
  CliUx.ux.action.start('Writing new version to konfig.yaml')
  fs.writeFileSync(konfigYamlPath, generateKonfigYamlString(loadedKonfigYaml))
  CliUx.ux.action.stop()
}
