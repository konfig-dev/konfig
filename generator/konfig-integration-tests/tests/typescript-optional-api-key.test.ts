import { e2e } from "../util";

test("typescript-optional-api-key", async () => {
  await e2e(4004);
});