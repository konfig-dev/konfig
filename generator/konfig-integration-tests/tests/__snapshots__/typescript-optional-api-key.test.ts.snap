// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typescript-optional-api-key 1`] = `
"# typescript-optional-api-key<a id=\\"typescript-optional-api-key\\"></a>

A simple API based for testing typescript-optional-api-key.

[![npm](https://img.shields.io/badge/npm-v1.0.0-blue)](https://www.npmjs.com/package/typescript-optional-api-key/v/1.0.0)
[![More Info](https://img.shields.io/badge/More%20Info-Click%20Here-orange)](http://example.com/support)

## Table of Contents<a id=\\"table-of-contents\\"></a>

<!-- toc -->

- [Installing](#installing)
  * [npm](#npm)
  * [yarn](#yarn)
- [Getting Started](#getting-started)
- [Reference](#reference)
  * [\`typescriptoptionalapikey.test.fetch\`](#typescriptoptionalapikeytestfetch)
  * [\`typescriptoptionalapikey.test.fetchNoSecurity\`](#typescriptoptionalapikeytestfetchnosecurity)

<!-- tocstop -->

## Installing<a id=\\"installing\\"></a>

### npm<a id=\\"npm\\"></a>
\`\`\`
npm install typescript-optional-api-key --save
\`\`\`

### yarn<a id=\\"yarn\\"></a>
\`\`\`
yarn add typescript-optional-api-key
\`\`\`

## Getting Started<a id=\\"getting-started\\"></a>

\`\`\`typescript
import { TypescriptOptionalApiKey } from \\"typescript-optional-api-key\\";

const typescriptoptionalapikey = new TypescriptOptionalApiKey({
  // Defining the base path is optional and defaults to https://typescript-optional-api-key.konfigthis.com
  // basePath: \\"https://typescript-optional-api-key.konfigthis.com\\",
  apiKey: \\"API_KEY\\",
});

const fetchResponse = await typescriptoptionalapikey.test.fetch();

console.log(fetchResponse);
\`\`\`

## Reference<a id=\\"reference\\"></a>


### \`typescriptoptionalapikey.test.fetch\`<a id=\\"typescriptoptionalapikeytestfetch\\"></a>

Provide an input parameter to receive a JSON value with properties.

#### üõ†Ô∏è Usage<a id=\\"üõ†Ô∏è-usage\\"></a>

\`\`\`typescript
const fetchResponse = await typescriptoptionalapikey.test.fetch();
\`\`\`

#### üåê Endpoint<a id=\\"üåê-endpoint\\"></a>

\`/simple-endpoint\` \`GET\`

[üîô **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptoptionalapikey.test.fetchNoSecurity\`<a id=\\"typescriptoptionalapikeytestfetchnosecurity\\"></a>

Provide an input parameter to receive a JSON value with properties.

#### üõ†Ô∏è Usage<a id=\\"üõ†Ô∏è-usage\\"></a>

\`\`\`typescript
const fetchNoSecurityResponse =
  await typescriptoptionalapikey.test.fetchNoSecurity();
\`\`\`

#### üåê Endpoint<a id=\\"üåê-endpoint\\"></a>

\`/endpoint-with-no-security\` \`GET\`

[üîô **Back to Table of Contents**](#table-of-contents)

---


## Author<a id=\\"author\\"></a>
This TypeScript package is automatically generated by [Konfig](https://konfigthis.com)
"
`;
